### PR TITLE
feat(ui): add new table confirm component

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Add a new table confirm component. 
- Update the domain table to use the new component. 
- Domain improvements from review.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to /r/domains.
- Check that the domain confirmation has a white background.
- Check that you can change the default domain.

Screen Shot 2021-06-11 at 4 30 19 pm

### CODE REVIEW

File: ui/src/app/pools/views/Pools.js [hunk 1]
```diff
@@ -69,7 +69,7 @@ const generateRows = (
                   "Cannot delete a pool that contains machines.")
               }
               editDisabled={!row.permissions.includes("edit")}
-              editPath={poolsURLs.edit({ edit: row.id })}
+              editPath={poolsURLs.edit({ id: row.id })}
```
Comment (huwshimi) [inline ui/src/app/pools/views/Pools.js:72]: This is just a driveby fix for going to the correct url.

File: ui/src/app/base/components/TableConfirm/TableConfirm.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  errorKey?: string;
+  finished: boolean;
+  inProgress: boolean;
+  message: ReactNode;
+  onClose: () => void;
+  onConfirm: () => void;
+  sidebar?: boolean;
+};
+
+const TableConfirm = ({
```
Comment (huwshimi) [inline ui/src/app/base/components/TableConfirm/TableConfirm.tsx:28]: @ClementChaumel here's the new confirm component I added that fits in with the style of our delete confirmations and I also updated it to handle API errors.

File: ui/src/app/domains/views/DomainsList/DomainsTable/DomainsTable.tsx [hunk 3]
```diff
@@ -113,6 +129,7 @@ const DomainsTable = (): JSX.Element => {
 
   return (
     <MainTable
+      className="p-table-expanding--light"
```
Comment (huwshimi) [inline ui/src/app/domains/views/DomainsList/DomainsTable/DomainsTable.tsx:132]: @ClementChaumel I figured out that this class was required for the confirmation to have the white background.

File: ui/src/app/store/domain/slice.ts [hunk 4]
```diff
@@ -38,10 +43,19 @@ const domainSlice = createSlice({
...
+    setDefaultError: (
+      state: DomainState,
+      action: PayloadAction<SetDefaultErrors>
+    ) => {
       state.saving = false;
       // API seems to return the domain id in payload.error not an error message
-      state.errors = "There was an error when setting default domain.";
+      // when the domain can't be found. This override can be removed when the
+      // bug is fixed: https://bugs.launchpad.net/maas/+bug/1931654.
+      if (typeof action.payload === "number") {
```
Comment (huwshimi) [inline ui/src/app/store/domain/slice.ts:54]: @ClementChaumel here I'm only overriding API errors when the error is a number so that the other errors are passed through.

File: ui/src/app/domains/views/DomainsList/DomainsTable/DomainsTable.tsx (hunk 3 above)
Comment (ClementChaumel) [reply ui/src/app/domains/views/DomainsList/DomainsTable/DomainsTable.tsx:132]: Nice one!

File: ui/src/app/base/components/TableConfirm/TableConfirm.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+} from "@canonical/react-components";
+import type { Props as ButtonProps } from "@canonical/react-components/dist/components/Button";
+
+import { COL_SIZES } from "app/base/constants";
+import { useCycled } from "app/base/hooks";
+
+export type Props = {
+  confirmAppearance?: ButtonProps["appearance"];
+  confirmLabel: string;
+  errors?: string | Record<string, string[]>;
```
Comment (Caleb-Ellis) [inline ui/src/app/base/components/TableConfirm/TableConfirm.tsx:18]: Should we be using the `FormErrors` type here? Maybe it needs to go in base types

File: ui/src/app/pools/views/Pools.js (hunk 1 above)
Comment (Caleb-Ellis) [reply ui/src/app/pools/views/Pools.js:72]: whoops, how did that happen? :grimacing:

**Review by Caleb-Ellis — APPROVED** [review]
Nice, I think this works well.

I think later down the line we could consider using `FormikForm` as the base for all of our forms, and something like this new `TableConfirm` component would wrap around it, setting some reasonable prop defaults and maybe adding an extra layer of functionality. But at least at the core things like error handling and validation and analytics are all done the same way.

File: ui/src/app/pools/views/Pools.js (hunk 1 above)
Comment (huwshimi) [reply ui/src/app/pools/views/Pools.js:72]: This is what happens when it's not a TypeScript file!

File: ui/src/app/base/components/TableConfirm/TableConfirm.tsx (hunk 5 above)
Comment (huwshimi) [reply ui/src/app/base/components/TableConfirm/TableConfirm.tsx:18]: Hmm yeah maybe, or maybe the `FormErrors` type should eventually become a generic type for the API error responses (kind of like I did for the setDefaultErrors case).
